src/devices/rpLidar2/sdk_original/sdk/src/rplidar_driver.cpp

```diff
@@ -0,0 +1,1026 @@
+            } else {
+                // current data is invalid, do not use it.
+                continue;
+            }
+        }
+
+        _capsuleToNormal(capsule_node, local_buf, count);
+
+        for (size_t pos = 0; pos < count; ++pos)
+        {
+            if (local_buf[pos].sync_quality & RPLIDAR_RESP_MEASUREMENT_SYNCBIT)
+            {
+                // only publish the data when it contains a full 360 degree scan 
+                
+                if ((local_scan[0].sync_quality & RPLIDAR_RESP_MEASUREMENT_SYNCBIT)) {
+                    _lock.lock();
+                    memcpy(_cached_scan_node_buf, local_scan, scan_count*sizeof(rplidar_response_measurement_node_t));
+                    _cached_scan_node_count = scan_count;
+                    _dataEvt.set();
+                    _lock.unlock();
+                }
+                scan_count = 0;
+            }
+            local_scan[scan_count++] = local_buf[pos];
+            if (scan_count == _countof(local_scan)) scan_count-=1; // prevent overflow
+        }
+    }
+    _isScanning = false;
+
+    return RESULT_OK;
+}
+
+u_result RPlidarDriverSerialImpl::grabScanData(rplidar_response_measurement_node_t * nodebuffer, size_t & count, _u32 timeout)
+{
+    switch (_dataEvt.wait(timeout))
+    {
+    case rp::hal::Event::EVENT_TIMEOUT:
+        count = 0;
+        return RESULT_OPERATION_TIMEOUT;
+    case rp::hal::Event::EVENT_OK:
+        {
+            if(_cached_scan_node_count == 0) return RESULT_OPERATION_TIMEOUT; //consider as timeout
+
+            rp::hal::AutoLocker l(_lock);
+
+            size_t size_to_copy = min(count, _cached_scan_node_count);
+
+            memcpy(nodebuffer, _cached_scan_node_buf, size_to_copy*sizeof(rplidar_response_measurement_node_t));
+            count = size_to_copy;
+            _cached_scan_node_count = 0;
+        }
+        return RESULT_OK;
+
+    default:
+        count = 0;
+        return RESULT_OPERATION_FAIL;
+    }
+}
+
+u_result RPlidarDriverSerialImpl::ascendScanData(rplidar_response_measurement_node_t * nodebuffer, size_t count)
+{
+    float inc_origin_angle = 360.0/count;
+    size_t i = 0;
+
+    //Tune head
+    for (i = 0; i < count; i++) {
+        if(nodebuffer[i].distance_q2 == 0) {
+            continue;
+        } else {
+            while(i != 0) {
+                i--;
+                float expect_angle = (nodebuffer[i+1].angle_q6_checkbit >> RPLIDAR_RESP_MEASUREMENT_ANGLE_SHIFT)/64.0f - inc_origin_angle;
+                if (expect_angle < 0.0f) expect_angle = 0.0f;
+                _u16 checkbit = nodebuffer[i].angle_q6_checkbit & RPLIDAR_RESP_MEASUREMENT_CHECKBIT;
+                nodebuffer[i].angle_q6_checkbit = (((_u16)(expect_angle * 64.0f)) << RPLIDAR_RESP_MEASUREMENT_ANGLE_SHIFT) + checkbit;
+            }
+            break;
+        }
+    }
+
+    // all the data is invalid
+    if (i == count) return RESULT_OPERATION_FAIL;
+
+    //Tune tail
+    for (i = count - 1; i >= 0; i--) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unsigned expression 'i' can't be negative so it is unnecessary to test it.](https://app.codacy.com/manual/robotology/yarp/pullRequest?prid=5162834)